room/compiler-processing/src/test/java/androidx/room/compiler/processing/XAnnotationTest.kt

```diff
@@ -0,0 +1,709 @@
+
+            @MainAnnotation(
+                typeList = [String::class, Int::class],
+                singleType = Long::class,
+                intMethod = 3,
+                otherAnnotationArray = [
+                    OtherAnnotation(
+                        value = "other list 1"
+                    ),
+                    OtherAnnotation(
+                        value = "other list 2"
+                    )
+                ],
+                singleOtherAnnotation = OtherAnnotation("other single")
+            )
+            public class Subject {
+            }
+            """.trimIndent()
+        )
+        runTest(
+            listOf(mySource)
+        ) { invocation ->
+            val element = invocation.processingEnv.requireTypeElement("Subject")
+            val annotation = element.requireAnnotation<MainAnnotation>()
+
+            assertThat(
+                annotation.get<List<XType>>("typeList").map {
+                    it.typeName
+                }
+            ).containsExactly(
+                String::class.typeName(),
+                Int::class.typeName()
+            )
+            assertThat(
+                annotation.get<XType>("singleType")
+            ).isEqualTo(
+                invocation.processingEnv.requireType(Long::class.typeName())
+            )
+
+            assertThat(annotation.get<Int>("intMethod")).isEqualTo(3)
+            annotation.get<XAnnotation>("singleOtherAnnotation")
+                .let { other ->
+                    assertThat(other.simpleName).isEqualTo("OtherAnnotation")
+                    assertThat(other.get<String>("value"))
+                        .isEqualTo("other single")
+                }
+            annotation.get<List<XAnnotation>>("otherAnnotationArray")
+                .let { boxArray ->
+                    assertThat(boxArray).hasSize(2)
+                    assertThat(boxArray[0].get<String>("value")).isEqualTo("other list 1")
+                    assertThat(boxArray[1].get<String>("value")).isEqualTo("other list 2")
+                }
+        }
+    }
+
+    @Test
+    fun typeReferenceArray_singleItemInJava() {
+        val src = Source.java(
+            "Subject",
+            """
+            import androidx.room.compiler.processing.testcode.JavaAnnotationWithTypeReferences;
+            @JavaAnnotationWithTypeReferences(String.class)
+            class Subject {
+            }
+            """.trimIndent()
+        )
+        runTest(
+            sources = listOf(src)
+        ) { invocation ->
+            val subject = invocation.processingEnv.requireTypeElement("Subject")
+            val annotation = subject.requireAnnotation<JavaAnnotationWithTypeReferences>()
+            val annotationValue = if (invocation.isKsp && !preCompiled) {
+                // TODO: 5/24/21 KSP does not wrap a single item in a list, even though the
+                // return type should be Class<?>[] (only in sources).
```

Comment:
could we not handle this by loading the annotation class from resolver and see if the method is declared as a list? 
i assume this might be happening in other types too?

Reply:
Good idea, I think I have a good general workaround for this now